### PR TITLE
Content Pack for Fortanix DSM Integration

### PR DESCRIPTION
**Pull request by arvin-dt**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Introduced an integration between PaloAlto/XSOAR/Demisto with Fortanix DSM. This allows customers to fetch secrets and perform cryptographic operations using a NIST-certified FIPS 140-2 Level 3 HSM.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0
- [x] 6.6.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Review by dansterenson — CHANGES_REQUESTED** [review]
Thanks! see my notes. Please fix them and let's schedule a demo to see this integration in action.

**Comment by arvin-dt** [comment]
> Thanks! see my notes. Please fix them and let's schedule a demo to see this integration in action.

Here's a [recorded demo](https://github.com/fortanix/Cortex-XSOAR-integration-Fortanix-DSM/blob/test/docs/Demo-20221031-Cortex-XSOAR-integration-Fortanix-DSM.mp4).

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.yml [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  name: token
+  type: 4
+  additionalinfo: Private key cannot be encrypted
+  required: false
+- display: Trust any server certificate (insecure)
+  name: tls_enforce
+  defaultvalue: "true"
+  type: 8
+  additionalinfo: Ignores TLS, not recommended
+  required: false
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.yml]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+
+def main() -> None:
+
+    message = ''
+    args = demisto.args()
+    dparams = demisto.params()
+    command = demisto.command()
+
+    proxy = dparams.get('proxy', False)
+    verify_ssl = not dparams.get('tls_enforce', False)
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+            k_v = attr.strip().split('=')
+            metadata[k_v[0]] = k_v[1]
+    return metadata
+
+
+def get_headers(username, password, apikey=None, bearer=None):
+    headers = {'Content-Type': 'application/json'}
+
+    if bearer and bearer != '':
+        headers['Authorization'] = 'Bearer ' + bearer
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+    if username and password and username != '' and password != '':
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+        if username.find('CERTIFICATE') < 0 and password.find('PRIVATE KEY') < 0:
+            headers['Authorization'] = 'Basic ' + \
+                b64encode(b":".join([username.encode("latin1"),
+                                     password.encode("latin1")])).decode('ascii')
+    elif apikey and apikey != '':
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+        headers['Authorization'] = 'Basic ' + apikey
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 7]
```diff
@@ -0,0 +1,533 @@
...
+
+        elif command == 'fortanix-decrypt':
+            return_results(decrypt_command(client, args, pkey, pmode))
+
+    except DemistoException as err:
+        message = f'Failure in Fortanix DSM.\nError:\n{str(err)}'
+    except ValueError as err:
+        message = f'Error in Fortanix DSM.\nError:\n{str(err)}'
+    except Exception as e:
+        message = f'Command Failure: {command}.\nError:\n{str(e)}'
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+
+    if message != '':
+        return_error(message)
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+
+def getMetadata(strMeta):
+    metadata = {}
+    try:
+        if strMeta:
+            if isJSON(strMeta):
+                metadata = json.loads(strMeta)
+    except Exception:
+        demisto.debug('Metadata is null')
+    if metadata is None or len(metadata) == 0:
```
[the remark this answers is not in the document]
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+        params['name'] = name
+    else:
+        if group_id := args.get('group_id'):
+            params['group_id'] = group_id
+        elif gids and len(gids) and gids.find(' ') < 0 and gids.find(',') < 0:
+            params['group_id'] = gids
+
+    res = client.send_request(path, 'get', params=params)
+    if not res:
+        raise DemistoException('Secrets not found')
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:142]: When there are no secrets found, an empty list is returned from the API? If so, it is better to return an empty list and not throw an exception. It is of if the list is empty, it should not fail the playbook or flow.

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+
+    if kid and res:
+        readable_output = tableToMarkdown(f'Fortanix DSM Secret {res}', res)
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:145]: Why {res} is in the table name?

```suggestion
        readable_output = tableToMarkdown(f'Fortanix DSM Secret', res)
```

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+        return CommandResults(
+            outputs_prefix='Fortanix.Secret',
+            readable_output=readable_output,
+            outputs=res,
+            raw_response=res
+        )
+    elif len(res) == 1:
+        readable_output = tableToMarkdown(f'Fortanix DSM Secret {res[0]}', res[0])
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:153]: same
```suggestion
        readable_output = tableToMarkdown(f'Fortanix DSM Secret', res[0])
```

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+                    value = b64decode(res['value']).decode("utf-8")
+                else:
+                    value = res['value']
+        except Exception:
+            value = res['value']
+
+    mapped_value = [{
+        'Value': value
+    }]
+    readable_output = tableToMarkdown(f'Fortanix DSM Secret {mapped_value}', mapped_value)
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:204]:
```suggestion
    readable_output = tableToMarkdown(f'Fortanix DSM Secret', mapped_value)
```

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+    if metadata := getMetadata(args.get('metadata')):
+        body['custom_metadata'] = metadata
+
+    res = client.send_request(path, method, body)
+    if not res:
+        msg = 'Secret cannot be '
+        msg = msg + 'created.' if rotate else msg + 'rotated.'
+        raise DemistoException(msg)
+
+    readable_output = tableToMarkdown(f'Fortanix DSM Secret {res}', res)
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:245]:
```suggestion
    readable_output = tableToMarkdown(f'Fortanix DSM Secret', res)
```

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 15]
```diff
@@ -0,0 +1,525 @@
...
+
+    path = 'crypto/v1/keys/' + kid
+    res = client.send_request(path, "delete")
+    if res:
+        raise DemistoException('Secret was not deleted.')
+
+    mapped_value = [{
+        'Result': 'OK'
+    }]
+    readable_output = tableToMarkdown(f'Fortanix DSM Secret {mapped_value}', mapped_value)
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:267]:
```suggestion
    readable_output = tableToMarkdown(f'Fortanix DSM Secret', mapped_value)
```

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+    if res and isinstance(res, str):
+        value = [{
+            'Output': res
+        }]
+    else:
+        value = [{
+            'Output': 'OK'
+        }]
+
+    readable_output = tableToMarkdown(f'Fortanix DSM Plugin {value}', value)
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:307]:
```suggestion
    readable_output = tableToMarkdown(f'Fortanix DSM Plugin', value)
```

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 17]
```diff
@@ -0,0 +1,525 @@
...
+            value = b64encode(json.dumps(res).encode('ascii')).decode('ascii')
+        else:
+            value = res
+    except Exception:
+        value = res
+
+    mapped_value = [{
+        'Cipher': value
+    }]
+    readable_output = tableToMarkdown(f'Fortanix DSM Encryption {mapped_value}', mapped_value)
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:352]:
```suggestion
    readable_output = tableToMarkdown(f'Fortanix DSM Encryption', mapped_value)
```

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py [hunk 18]
```diff
@@ -0,0 +1,525 @@
...
+                value = res['plain']
+        else:
+            value = res
+    except Exception:
+        value = res['plain']
+
+    mapped_value = [{
+        'Plain': value
+    }]
+    readable_output = tableToMarkdown(f'Fortanix DSM Decryption {mapped_value}', mapped_value)
```
Comment (dansterenson) [inline Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py:437]:
```suggestion
    readable_output = tableToMarkdown(f'Fortanix DSM Decryption', mapped_value)
```

**Review by dansterenson — CHANGES_REQUESTED** [review]
I watched the demo, well done! See my small notes. 
Please add some Uni Tests to your helper functions and command functions. You can see some good examples [here](https://github.com/demisto/content/tree/master/Packs/HelloWorld/Integrations/HelloWorld) under **HelloWorld_test.py**.

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 10 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 11 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 12 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 13 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 15 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 16 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 17 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

File: Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py (hunk 18 above)
Comment (arvin-dt) [reply Packs/Fortanix-DSM/Integrations/FortanixDSM/FortanixDSM.py]: done

**Pull request merged**